### PR TITLE
Add diagnostic and hover language server

### PR DESCRIPTION
**Pull request by ayazhafiz**
This is a combination of 5 commits.

- Basic server implementation
- Implement diagnostic publishing
- Add support for hover
- Add tests for diagnostics
- Add tests for hover

### CODE REVIEW

**Comment by ayazhafiz** [comment]
@lukebhan - can you take a look. Do you want me to add tests?

File: editor/language_server/src/diagnostics.rs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        .collect()
+}
+
+fn to_severity(diagnostic_kind: &s::DiagnosticKind) -> DiagnosticSeverity {
+    match diagnostic_kind {
+        s::DiagnosticKind::Error => DiagnosticSeverity::Error,
+        s::DiagnosticKind::Warning => DiagnosticSeverity::Warning,
+        s::DiagnosticKind::Note => DiagnosticSeverity::Information,
+        s::DiagnosticKind::Help => DiagnosticSeverity::Hint,
+    }
```
Comment (lukebhan) [inline editor/language_server/src/diagnostics.rs:56]: this seems useless. why dont you just replace severity with kind in a duagnostic above. also to/from would be better if not

File: editor/language_server/src/lsp.rs [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    async fn initialized(&self, _params: InitializedParams) {
+        self.client
+            .log_message(MessageType::Info, "Slide language server initialized.")
+            .await;
+    }
+
+    async fn shutdown(&self) -> Result<()> {
+        Ok(())
```
Comment (lukebhan) [inline editor/language_server/src/lsp.rs:107]: does nothing?

File: editor/language_server/src/lsp.rs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    async fn did_open(&self, params: DidOpenTextDocumentParams) {
+        let TextDocumentItem { uri, text, version, .. } = params.text_document;
+        self.change(uri, text, Some(version)).await;
+    }
```
Comment (lukebhan) [inline editor/language_server/src/lsp.rs:113]: shouldnt this return something just or i guess its async so maybe not.

**Review by lukebhan — APPROVED** [review]

**Comment by lukebhan** [comment]
> @lukebhan - can you take a look. Do you want me to add tests?

you probably should but i trust it work. i added some thins i noticed.

File: editor/language_server/src/diagnostics.rs (hunk 1 above)
Comment (ayazhafiz) [reply editor/language_server/src/diagnostics.rs:56]: I can't, I have to change the slide diagnostic type to work with the other diagnostic type, which expects this other enum

File: editor/language_server/src/lsp.rs (hunk 2 above)
Comment (ayazhafiz) [reply editor/language_server/src/lsp.rs:107]: Yep, we just have to implement it

File: editor/language_server/src/lsp.rs (hunk 3 above)
Comment (ayazhafiz) [reply editor/language_server/src/lsp.rs:113]: yeah nothing to ret

**Pull request merged**